File: Design-Patterns-Using-Cpp/ChainOfResponsibility/ChainOfResponsibilityGUI.cpp
#include <iostream>
#include <string>

class Handler {
protected:
    Handler* parent;
    std::string helpText;

public:
    Handler(Handler* parent) : parent(parent), helpText("") {}

    virtual void showHelperText() = 0;

    void setHelperText(const std::string& text) {
        helpText = text;
    }
};

class Container : public Handler {
public:
    Container(Handler* parent) : Handler(parent) {}

    void showHelperText() override {
        if (!helpText.empty()) {
            std::cout << "Help :: " << helpText << std::endl;
        } else if (parent != nullptr) {
            std::cout << "Message passed to next in chain by Container" << std::endl;
            parent->showHelperText();
        }
    }
};

class Button : public Handler {
private:
    std::string label;

public:
    Button(const std::string& label, Handler* parent) : Handler(parent), label(label) {}

    void showHelperText() override {
        if (!helpText.empty()) {
            std::cout << "Help :: " << helpText << std::endl;
        } else if (parent != nullptr) {
            std::cout << "Message passed to next in chain by Button" << std::endl;
            parent->showHelperText();
        }
    }
};

class Panel : public Handler {
public:
    Panel() : Handler(nullptr) {}

    void showHelperText() override {
        if (!helpText.empty()) {
            std::cout << "Help :: " << helpText << std::endl;
        } else if (parent != nullptr) {
            std::cout << "Message passed to next in chain by Panel" << std::endl;
            parent->showHelperText();
        }
    }
};

// Client code
int main() {
    Panel p;
    p.setHelperText("Panel help text.");

    Button b1("Ok", &p);
    b1.setHelperText("Ok button help text.");

    Button b2("Cancel", &p);

    b1.showHelperText();
    b2.showHelperText();

    return 0;
}

/*
Help :: Ok button help text.
Message passed to next in chain by Button
Help :: Panel help text.
*/

File: Design-Patterns-Using-Go/ChainOfResponsibility/ChainOfResponsibilityGUI.go
package main

import "fmt"

// Handler interface
type Handler interface {
	showHelperText()
	setHelperText(text string)
}

// Container struct
type Container struct {
	parent   Handler
	helpText string
}

// NewContainer constructor
func NewContainer(parent Handler) *Container {
	return &Container{
		parent: parent,
	}
}

// showHelperText method for Container
func (c *Container) showHelperText() {
	if c.helpText != "" {
		fmt.Println("Help ::", c.helpText)
	} else if c.parent != nil {
		fmt.Println("Message passed to next in chain by Container")
		c.parent.showHelperText()
	}
}

// setHelperText method for Container
func (c *Container) setHelperText(text string) {
	c.helpText = text
}

// Button struct
type Button struct {
	parent   Handler
	label    string
	helpText string
}

// NewButton constructor
func NewButton(label string, parent Handler) *Button {
	return &Button{
		parent: parent,
		label:  label,
	}
}

// showHelperText method for Button
func (b *Button) showHelperText() {
	if b.helpText != "" {
		fmt.Println("Help ::", b.helpText)
	} else if b.parent != nil {
		fmt.Println("Message passed to next in chain by Button")
		b.parent.showHelperText()
	}
}

// setHelperText method for Button
func (b *Button) setHelperText(text string) {
	b.helpText = text
}

// Panel struct
type Panel struct {
	parent   Handler
	helpText string
}

// NewPanel constructor
func NewPanel(parent Handler) *Panel {
	return &Panel{
		parent: parent,
	}
}

// showHelperText method for Panel
func (p *Panel) showHelperText() {
	if p.helpText != "" {
		fmt.Println("Help ::", p.helpText)
	} else if p.parent != nil {
		fmt.Println("Message passed to next in chain by Panel")
		p.parent.showHelperText()
	}
}

// setHelperText method for Panel
func (p *Panel) setHelperText(text string) {
	p.helpText = text
}

// Client Code
func main() {
	p := NewPanel(nil)
	p.setHelperText("Panel help text.")

	b1 := NewButton("Ok", p)
	b1.setHelperText("Ok button help text.")
	b2 := NewButton("Cancel", p)

	b1.showHelperText()
	b2.showHelperText()
}

/*
Help :: Ok button help text.
Message passed to next in chain by Button
Help :: Panel help text.
*/